util/rbac/rbac.go

```diff
@@ -426,20 +463,81 @@ func PolicyCSV(data map[string]string) string {
 			strBuilder.WriteString(value)
 		}
 	}
+}
+
+// PolicyCSV will generate the final policy csv to be used
+// by Argo CD RBAC from a single configmap. It will find entries
+// in the given data that matches the policy key name convention:
+//
+//	policy[.overlay].csv
+func PolicyCSV(data map[string]string) string {
+	var strBuilder strings.Builder
+	policyCSV(&strBuilder, data)
+	return strBuilder.String()
+}
+
+// PolicyCSV will generate the final policy csv to be used
+// by Argo CD RBAC from many configmaps. It will find entries
+// in the given data that matches the policy key name convention:
+//
+//	policy[.overlay].csv
+func PolicyCSVMany(datas []map[string]string) string {
+	var strBuilder strings.Builder
+	// add the main policy first
+	first := true
+	for _, data := range datas {
+		if _, ok := data[ConfigMapPolicyCSVKey]; ok && !first {
+			strBuilder.WriteString("\n")
+		}
+		first = false
+
+		policyCSV(&strBuilder, data)
+	}
 	return strBuilder.String()
 }
 
-// syncUpdate updates the enforcer
-func (e *Enforcer) syncUpdate(cm *apiv1.ConfigMap, onUpdated func(cm *apiv1.ConfigMap) error) error {
-	e.SetDefaultRole(cm.Data[ConfigMapPolicyDefaultKey])
-	e.SetMatchMode(cm.Data[ConfigMapMatchModeKey])
-	policyCSV := PolicyCSV(cm.Data)
+// sync updates the enforcer after a configmap has been upserted or removed
+func (e *Enforcer) sync(cm *apiv1.ConfigMap, onUpdated func(cm *apiv1.ConfigMap) error) error {
+	cmNames := make([]string, 0, len(e.policyCSVCache))
+	for key := range e.policyCSVCache {
+		cmNames = append(cmNames, key)
+	}
+	slices.Sort(cmNames)
+	datas := make([]map[string]string, len(cmNames))
+	for ix, name := range cmNames {
+		datas[ix] = e.policyCSVCache[name]
+	}
+
+	policyCSV := PolicyCSVMany(datas)
 	if err := onUpdated(cm); err != nil {
 		return err
 	}
 	return e.SetUserPolicy(policyCSV)
 }
 
+// syncUpdate updates the enforcer to upsert a configmap
+func (e *Enforcer) syncUpdate(cm *apiv1.ConfigMap, onUpdated func(cm *apiv1.ConfigMap) error) error {
+	if cm.Name == e.configmap {
+		e.SetDefaultRole(cm.Data[ConfigMapPolicyDefaultKey])
+		e.SetMatchMode(cm.Data[ConfigMapMatchModeKey])
+	}
+	e.policyCSVCache[cm.Name] = cm.Data
+
+	return e.sync(cm, onUpdated)
+}
+
+// syncUpdate updates the enforcer to remove a configmap
+func (e *Enforcer) syncRemove(cm *apiv1.ConfigMap, onUpdated func(cm *apiv1.ConfigMap) error) error {
+	if cm.Name == e.configmap {
+		// This should never happen?
+		// Assume this is user error and ignore it
+		return nil
```

Comment:
Why do we ignore it instead of surfacing?

Reply:
Not sure, wrote this over half a year ago, so my memory is hazy at best. I think the logic was that this CM should never be deleted, so if it is, we shouldn't act on it. Or are you just suggesting to return an error so that it get's logged?

Reply:
Yeah, I'm thinking of at least logging the error.